eth/db/header.py

```diff
@@ -319,6 +320,8 @@ def _persist_header_chain(
                 rlp.encode(curr_chain_head),
             )
 
+            cls._add_block_number_to_hash_lookup(db, curr_chain_head)
```

Comment:
@carver Can I get your opinion on this?

I discovered a problem that we currently have in Py-EVM. Say you write header `10` as checkpoint header leading to a gap from `1` to `9`. At some point later I go and fetch the headers 1, 2, 3 from a peer with the intention to start closing the gap. Now the problem is, how do we know that the headers 1, 2, 3 do actually belong to the *canonical* chain? The call to `persist_header_chain` does validate the headers in the sense that they match the previous header (genesis, in this example) but it does not know if it will eventually lead us back to the header `10` that we accepted with a trusted score. It could also be an alternative history that we are accepting here. Eventually, as we are closing the gap at some point we will know if the headers match up to our expected header `10` but there is a window of uncertainty where we can't be exactly sure about it I believe.

I have two ideas how to go about that.

1. We could not assign block numbers until we finally closed the gap and then patch up the block numbers backwards from the closing point until we don't hit a missing one anymore. But that would mean that we can not read headers by block number while we are still working on closing the gap. We could still track the gap with the `header_db.get_header_chain_gaps()` API though.

2. We could add a safeguard that goes roughly like: If we are writing headers with a block number lower than the one of our current head and we do not have block numbers assigned to any headers for these block numbers yet, accept them as canonical headers even though we can not entirely be sure that these will eventually lead to our previously accepted trusted score header.

Reply:
I believe that the chain rules should supercede a checkpointed header.  In the case that we checkpoint at height 10, and then fill in 1-9 to find out that 9 is **not** the parent of 10, and that there is an alternate header at height 10 that would be picked by the fork choice rule we should discard our non-canonical 10 and all headers onward and re-sync.

At present we may be in a position where if we detect this situation we need to simply crash the client.  In practice I believe that we are unlikely to end up in this scenario without the user doing something very odd.

Reply:
> At present we may be in a position where if we detect this situation we need to simply crash the client.

Yeah, I think this is the best path, at least in the short term. If the user asserts something that trinity finds to be false, we should crash.

That means we should be extra careful to help people not pick a checkpoint that gets uncled (it was already true, just a reminder).

One thing to note: it gives other clients a way to crash our client, if they know where our checkpoint is: they respond with an uncle at the last unknown header in the gap. So one way to avoid crashing hard is to treat the very last header as invalid if it's not a parent of the canonical block at the child's slot. We would have to print loud warnings, though. But having a client crash that is externally triggerable is a problem.

Reply:
Thank you both for your input, it was really helpful.

I was starring on the screen for while, puzzled how to come up with an API that:

- doesn't need a new explicit param such as `force_canonical`
- passes all our tests that ensure tracking of forks and the canonical chain works as it does prior to this PR
- doesn't result in lots of additional db reads to figure out if when it would be appropriate to assign a header a block number (making it part of the canonical chain) when we write headers that are patching gaps.

At some point it occurred to me that `_update_header_chain_gaps` could just include the information of not just what are the new gaps but also what action was taken to get there (e.g. a gap was shrinked).

We can return this info without any additional db read and so I came up with a rule that says we always assign the block number in cases where we know that we have patched a gap in the chain.

This passes all existing tests without incurring much overhead and I'm quite happy with the API. That said, we could also move it to Trinity if you don't feel comfortable introducing these changes to Py-EVM. But I guess it would involve monkey patching Py-EVM to do the tracking which I'm not sure is a good idea.

But anyway, I'm not yet finished with the PR yet. This is just an intermediate update.

>One thing to note: it gives other clients a way to crash our client, if they know where our checkpoint is: they respond with an uncle at the last unknown header in the gap. So one way to avoid crashing hard is to treat the very last header as invalid if it's not a parent of the canonical block at the child's slot. We would have to print loud warnings, though. But having a client crash that is externally triggerable is a problem.

Yeah, I think that part is easy to handle on the Trinity side. :+1: